contracts/javascore/xcall/src/main/java/foundation/icon/btp/xcall/CallServiceImpl.java

```diff
@@ -0,0 +1,462 @@
+                cleanupCallRequest(resSn);
+                break;
+            case CSMessageResponse.FAILURE:
+            case CSMessageResponse.IBC_ERROR:
+            default:
+                // emit rollback event
+                Context.require(req.getRollback() != null, "NoRollbackData");
+                req.setEnabled();
+                requests.set(resSn, req);
+                RollbackMessage(resSn);
+        }
+    }
+
+    @External(readonly = true)
+    public Address admin() {
+        return admin.getOrDefault(Context.getOwner());
+    }
+
+    @External
+    public void setAdmin(Address _address) {
+        onlyOwner();
+        admin.set(_address);
+    }
+
+    @External
+    public void setProtocolFeeHandler(@Optional Address _addr) {
+        checkCallerOrThrow(admin(), "OnlyAdmin");
+        feeHandler.set(_addr);
+        if (_addr != null) {
+            var accruedFees = Context.getBalance(Context.getAddress());
+            if (accruedFees.signum() > 0) {
+                Context.transfer(_addr, accruedFees);
+            }
+        }
+    }
+
+    @External(readonly = true)
+    public Address getProtocolFeeHandler() {
+        return feeHandler.get();
+    }
+
+    @External
+    public void setProtocolFee(BigInteger _value) {
+        checkCallerOrThrow(admin(), "OnlyAdmin");
+        Context.require(_value.signum() >= 0, "ValueShouldBePositive");
+        protocolFee.set(_value);
+    }
+
+    @External(readonly = true)
+    public BigInteger getProtocolFee() {
+        return protocolFee.getOrDefault(BigInteger.ZERO);
+    }
+
+    @External
+    public void onChanOpenInit(int order, String[] connectionHops, String portId, String channelId,
+            byte[] counterpartyPb, String version) {
+        onlyIBCHandler();
+        sourcePort.set(portId);
+        sourceChannel.set(channelId);
+        Counterparty counterparty = Counterparty.decode(counterpartyPb);
+        destinationChannel.set(counterparty.getChannelId());
+        destinationPort.set(counterparty.getPortId());
+        Context.println("onChanOpenInit");
+    }
+
+    @External
+    public void onChanOpenTry(int order, String[] connectionHops, String portId, String channelId,
+            byte[] counterpartyPb, String version, String counterpartyVersion) {
+        onlyIBCHandler();
+        sourcePort.set(portId);
+        sourceChannel.set(channelId);
+        Counterparty counterparty = Counterparty.decode(counterpartyPb);
+        destinationChannel.set(counterparty.getChannelId());
+        destinationPort.set(counterparty.getPortId());
+        Context.println("onChanOpenTry");
+    }
+
+    @External
+    public void onChanOpenAck(String portId, String channelId, String counterpartyChannelId,
+            String counterpartyVersion) {
+        onlyIBCHandler();
+        Context.require(portId.equals(sourcePort.get()), "port not matched");
+        Context.require(channelId.equals(sourceChannel.get()), "Channel not matched");
```

Comment:
Need to set destination channel here from counterpartyChannelId, the one given in INIT will be empty.

Reply:
Added on next commit.